src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -177,5 +209,338 @@ private void ThrowIfDisposed()
+            if (_entityType == EntityType.Queue)
+            {
+                return await GetQueueMetricsAsync();
+            }
+            else
+            {
+                return await GetTopicMetricsAsync();
+            }
+        }
+
+        public async Task<ServiceBusTriggerMetrics> GetMetricsAsync()
+        {
+            if (_entityType == EntityType.Queue)
+            {
+                return await GetQueueMetricsAsync();
+            }
+            else
+            {
+                return await GetTopicMetricsAsync();
+            }
+        }
+
+        public async Task<ServiceBusTriggerMetrics> GetQueueMetricsAsync()
+        {
+            Message message = null;
+            QueueDescription queue;
+            QueueRuntimeInfo queueRuntime;
+            long deadLetterCount = 0, activeMessageCount = 0;
+            int partitionCount = 0;
+            TimeSpan eventQueueTime = TimeSpan.Zero;
+            bool isUnauthorized = false;
+
+            try
+            {
+                InitializeMetricsClients();
+
+                // Peek the first message in the queue without removing it from the queue
+                message = await _metricsReceiver.PeekAsync();
+
+                queueRuntime = await _managementClient.GetQueueRuntimeInfoAsync(_entityPath);
+                activeMessageCount = queueRuntime.MessageCountDetails.ActiveMessageCount;
+                deadLetterCount = queueRuntime.MessageCountDetails.DeadLetterMessageCount;
+
+                // If partitioning is turned on, then Service Bus automatically partitions queues into 16 partitions
+                // See more information here: https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-partitioning#standard
+                queue = await _managementClient.GetQueueAsync(_entityPath);
+                if (queue.EnablePartitioning)
+                {
+                    partitionCount = 16;
+                }
+            }
+            catch (MessagingEntityNotFoundException)
+            {
+                _logger.LogWarning($"ServiceBus queue '{_entityPath}' was not found.");
+            }
+            catch (UnauthorizedAccessException) // When Manage claim is not used on Service Bus connection string
+            {
+                if (TimeToLogWarning())
+                {
+                    _logger.LogWarning($"Connection string does not have Manage claim for queue '{_entityPath}'. Failed to get queue description to derive queue length metrics. " +
+                        $"Falling back to using first message enqueued time.");
+                }
+
+                isUnauthorized = true;
+            }
+            catch (Exception e)
+            {
+                _logger.LogWarning($"Error querying for Service Bus queue scale status: {e.Message}");
+            }
+
+            return CompileTriggerMetrics(message, activeMessageCount, deadLetterCount, partitionCount, isUnauthorized, _isListeningOnDeadLetterQueue);
```

Comment:
CompileTriggerMetrics is being used partially to avoid code duplication between queues and topics and partially to allow for testing of that logic.

Reply:
nit: maybe name CreateTriggerMetrics?